107-tee-boot-attestation.md

```diff
@@ -0,0 +1,360 @@
+successful boot attestation that will be required for guest OS to boot or a
+keylime agent to start running. In-effect, this handler would essentially
+bootstrap a keylime agent running in a CVM, as the TEE boot environment would
+need to be validated before a keylime agent could begin running.
+
+## Motivation
+
+<!--
+This section is for explicitly listing the motivation, goals and non-goals of
+this enhancement.  Describe why the change is important and the benefits to users.
+-->
+
+As CVMs are deployed on untrusted systems, it is reasonable to assume that a
+CVM user would also like to take advantage of system integrity monitoring
+provided by keylime *in addition to* TEE technology. That is, it is reasonable
+to expect that keylime agents will be deployed on CVMs. With that, it is
+reasonable to include TEE attestation mechanisms for environments in which
+keylime will already be running for system integrity monitoring purposes.
+
+### Goals
+
+<!--
+List the specific goals of the enhancement.  What is it trying to achieve?  How will we
+know that this has succeeded?
+-->
+ * Extend keylime verifier to include support for TEE boot attestation.
+ * Extend TPM measurements to account for TEE boot attestation within keylime.
+ * Fully support confidential computing systems' need for boot *and* runtime
+   attestation within keylime.
+
+### Non-Goals
+
+<!--
+What is out of scope for this enhancement?  Listing non-goals helps to focus discussion
+and make progress.
+-->
+ * TODO
+
+## Proposal
+
+<!--
+This is where we get down to the specifics of what the proposal actually is.
+This should have enough detail that reviewers can understand exactly what
+you're proposing, but should not include things like API designs or
+implementation.  The "Design Details" section below is for the real
+nitty-gritty.
+-->
+
+### User Stories (optional)
+
+<!--
+Detail the things that people will be able to do if this enhancement is implemented.
+Include as much detail as possible so that people can understand the "how" of
+the system.  The goal here is to make this feel real for users without getting
+bogged down.
+-->
+
+Initially, the TEE handler will work in-tandem with [SVSM], which runs in
+a tenant's firmware and provides privileged operations for CVM guests. SVSM will
+handle TEE attestation before a guest OS boots. SVSM will be responsible for
+unlocking state (provided by the keylime verifier) needed to begin the guest OS
+boot process.
+
+#### Story 1
+ * User launches a confidential VM, pointing it to the URL of the keylime
+   verifier's TEE handler to communicate with for boot attestation.
+ * CVM begins running in SVSM. SVSM gathers all necessary information needed for
+   TEE attestation. Information is dependant on the specific TEE architecture
+   the guest is running on top of (ARM CCA, Intel TDX, AMD SEV-SNP, etc...).
+ * SVSM sends all attestation information to TEE handler.
+ * TEE handler uses information to verify the boot environment.
```

Comment:
A portion of this data will be architecture-dependent and a portion will not be.

The portion that will be arch-dependent is the attestation evidence itself. This is because each TEE's attestation report from its secure processor is formatted differently and contains different information. Different types of X509 certificates or related items may also be included based on the architecture.

The portion that will not be arch-dependent is the public key to encrypt the secret with. Since a client like SVSM runs in guest VM firmware, a network stack is not yet available to it. As such, we must use the host for networking. However, for privacy purposes we cannot allow the host to read the attestation secret. As such, SVSM will create a private key in guest (encrypted) memory, give the host the public components of the key, and expect keylime to encrypt the secret upon a successful attestation. This would allow the guest VM to gain access to the resource without the host being able to also decrypt it.